src/Native/Unix/System.Native/pal_io.cpp

```diff
@@ -1171,3 +1171,44 @@ extern "C" char* SystemNative_RealPath(const char* path)
     assert(path != nullptr);
     return realpath(path, nullptr);
 }
+
+static int32_t LockUnlock(intptr_t fd, off_t offset, off_t length, bool lock)
+{
+    if (offset < 0 || length < 0) 
+    {
+        errno = EINVAL;
+        return -1;
+    }
+    
+    struct flock lockArgs;
+    int32_t ret;
+    if (lock)
+    {
+        lockArgs.l_type = F_WRLCK;
+    }
+    else
+    {
+        lockArgs.l_type = F_UNLCK;
+    }
+    lockArgs.l_whence = SEEK_SET;
+    lockArgs.l_start = offset;
+    lockArgs.l_len = length;
+    
+    do 
+    {
+        ret = fcntl (ToFileDescriptor(fd), F_SETLK, &lockArgs);
+    } while (ret == -1 && errno == EINTR);
+
+    return ret;
+}
+
+extern "C" int32_t SystemNative_LockFileRegion(int fd, off_t offset, off_t length)
+{
+    return LockUnlock(fd, offset, length, true);
+}
+
+extern "C" int32_t SystemNative_UnlockFileRegion(int fd, off_t offset, off_t length)
```

Comment:
off_t is variable in size based on platform... check what we've done elsewhere, but I expect these should all be int64_t